### PR TITLE
Add ReferenceManager to azure-core

### PR DESCRIPTION
**Pull request by alzimmermsft**
This PR adds `ReferenceManager` to `azure-core` which tracks an object reference until it is phantom reachable. Upon the referent becoming phantom reachable a cleanup action is ran, any exceptions thrown during cleanup are caught and ignored.

### CODE REVIEW

File: sdk/core/azure-core/src/main/java/com/azure/core/util/ReferenceManager.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.util;
+
+import com.azure.core.util.implementation.ReferenceManagerImpl;
+
+/**
+ * This interface represents the managing references to {@link Object Objects} and providing the ability to run a
+ * cleaning operation once the object is no longer able to be reference.
```
Comment (JonathanGiles) [inline sdk/core/azure-core/src/main/java/com/azure/core/util/ReferenceManager.java:10]: nit: hard to parse

File: sdk/core/azure-core/src/main/java/com/azure/core/util/implementation/ReferenceManagerImpl.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.util.implementation;
```
Comment (JonathanGiles) [inline sdk/core/azure-core/src/main/java/com/azure/core/util/implementation/ReferenceManagerImpl.java:4]: I'm not a fan of multiple implementation packages - I'd rather have a com.azure.core.implementation.*.*.* style than this.

File: sdk/core/azure-core/src/main/java/com/azure/core/util/implementation/ReferenceManagerImpl.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+import com.azure.core.util.ReferenceManager;
+
+import java.lang.ref.PhantomReference;
+import java.lang.ref.ReferenceQueue;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Implementation of {@link ReferenceManager}.
```
Comment (JonathanGiles) [inline sdk/core/azure-core/src/main/java/com/azure/core/util/implementation/ReferenceManagerImpl.java:14]: It would pay to note in the docs here that this is for Java 8 and lower, and there exists one for 9+ in the appropriate MR-JAR location (otherwise I can see frustrated devs wondering why this code isn't running!)

File: sdk/core/azure-core/src/main/java/com/azure/core/util/implementation/ReferenceManagerImpl.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+     * Creates a new instance of {@link ReferenceManagerImpl}.
+     */
+    public ReferenceManagerImpl() {
+        this.queue = new ReferenceQueue<>();
+        this.cleanableReferenceList = new CleanableReference<>();
+
+        new CleanableReference<>(this, () -> { }, this);
+
+        Thread thread = new Thread(Thread.currentThread().getThreadGroup(), this,
+            BASE_THREAD_NAME + RESOURCE_MANAGER_THREAD_NUMBER.getAndIncrement());
```
Comment (JonathanGiles) [inline sdk/core/azure-core/src/main/java/com/azure/core/util/implementation/ReferenceManagerImpl.java:36]: In what context is the `ReferenceManagerImpl` constructor called more than once? It appears to be a static field on `ReferenceManager`, and so I'm unsure why we need the thread number incrementing?

File: sdk/core/azure-core/src/main/java/com/azure/core/util/implementation/ReferenceManagerImpl.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    private final ReferenceQueue<Object> queue;
+
+    /**
+     * Creates a new instance of {@link ReferenceManagerImpl}.
+     */
+    public ReferenceManagerImpl() {
+        this.queue = new ReferenceQueue<>();
+        this.cleanableReferenceList = new CleanableReference<>();
+
+        new CleanableReference<>(this, () -> { }, this);
```
Comment (JonathanGiles) [inline sdk/core/azure-core/src/main/java/com/azure/core/util/implementation/ReferenceManagerImpl.java:33]: I would recommend a comment for this line, because as it is it looks very odd. You should get into the habit of talking to yourself in code like this, to explain why it is this way.

File: sdk/core/azure-core/src/main/java/com/azure/core/util/implementation/ReferenceManagerImpl.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    public void register(Object object, Runnable cleanupAction) {
+        new CleanableReference<>(Objects.requireNonNull(object, "'object' cannot be null."),
+            Objects.requireNonNull(cleanupAction, "'cleanupAction' cannot be null."), this);
+    }
+
+    @Override
+    public void run() {
+        while (cleanableReferenceList.hasRemaining()) {
+            try {
+                CleanableReference<?> reference = (CleanableReference<?>) queue.remove(30000);
```
Comment (JonathanGiles) [inline sdk/core/azure-core/src/main/java/com/azure/core/util/implementation/ReferenceManagerImpl.java:51]: Comment about the 30000 being a timeout (so people don't assume you've got weird code removing the first 30000 items on every iteration)

File: sdk/core/azure-core/src/main/java/com/azure/core/util/implementation/ReferenceManagerImpl.java [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+import java.lang.ref.PhantomReference;
+import java.lang.ref.ReferenceQueue;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Implementation of {@link ReferenceManager}.
+ */
+public final class ReferenceManagerImpl implements ReferenceManager, Runnable {
```
Comment (JonathanGiles) [inline sdk/core/azure-core/src/main/java/com/azure/core/util/implementation/ReferenceManagerImpl.java:16]: I'm not always a super-big fan of having classes implement `Runnable` like this, because it tends to conflate some 'outer' functionality with some 'inner' functionality. It took me a bit of time to identify where the `run()` was being called - I instinctively looked into the `CleanableReference` code first, and started to wonder if there was a bug, before I realised the new Thread was being given `this` to call it. Often what I would prefer to do is this (after removing `Runnable` from the class and renaming `run` to something like `clearReferenceQueue`):

```java
Thread thread = new Thread(Thread.currentThread().getThreadGroup(), this::clearReferenceQueue,
            BASE_THREAD_NAME + RESOURCE_MANAGER_THREAD_NUMBER.getAndIncrement());
```

File: sdk/core/azure-core/src/main/java/com/azure/core/util/ReferenceManager.java (hunk 1 above)
Comment (alzimmermsft) [reply sdk/core/azure-core/src/main/java/com/azure/core/util/ReferenceManager.java:10]: Words weren't done well here... will re-write this

File: sdk/core/azure-core/src/main/java/com/azure/core/util/implementation/ReferenceManagerImpl.java (hunk 2 above)
Comment (alzimmermsft) [reply sdk/core/azure-core/src/main/java/com/azure/core/util/implementation/ReferenceManagerImpl.java:4]: Was just adding a missing `package-info` in this case, we should discuss cleaning this up more in the future as we also have other package paths to implementation in Core, ex `com.azure.core.implementation`.

File: sdk/core/azure-core/src/main/java/com/azure/core/util/implementation/ReferenceManagerImpl.java (hunk 4 above)
Comment (alzimmermsft) [reply sdk/core/azure-core/src/main/java/com/azure/core/util/implementation/ReferenceManagerImpl.java:36]: Hold over from when multiple instance could be created, I'll comment this out and simplify it

**Review by JonathanGiles — APPROVED** [review]

File: sdk/core/azure-core/src/main/java/com/azure/core/implementation/util/ReferenceManagerImpl.java [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+     */
+    public ReferenceManagerImpl() {
+        this.queue = new ReferenceQueue<>();
+        this.cleanableReferenceList = new CleanableReference<>();
+
+        // Register this instance of ReferenceManager as the head of the cleaning queue. Doing so will allow the
+        // ReferenceManager to clean itself up when no longer in use, for now this is a no-op.
+        new CleanableReference<>(this, () -> { }, this);
+
+        Thread thread = new Thread(Thread.currentThread().getThreadGroup(), this::clearReferenceQueue,
```
Comment (conniey) [inline sdk/core/azure-core/src/main/java/com/azure/core/implementation/util/ReferenceManagerImpl.java:43]: I'm curious why we don't hold a reference to it and join this when it ends? Feels weird to not reference a long-lived object.

File: sdk/core/azure-core/src/main/java/com/azure/core/implementation/util/ReferenceManagerImpl.java [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+    /**
+     * Creates a new instance of {@link ReferenceManagerImpl}.
+     */
+    public ReferenceManagerImpl() {
+        this.queue = new ReferenceQueue<>();
+        this.cleanableReferenceList = new CleanableReference<>();
+
+        // Register this instance of ReferenceManager as the head of the cleaning queue. Doing so will allow the
+        // ReferenceManager to clean itself up when no longer in use, for now this is a no-op.
+        new CleanableReference<>(this, () -> { }, this);
```
Comment (conniey) [inline sdk/core/azure-core/src/main/java/com/azure/core/implementation/util/ReferenceManagerImpl.java:41]: Generally not a fan of referencing itself. Creates a circular dependency... This class is a private static final. Do we need this class?

File: sdk/core/ci.yml [hunk 10]
```diff
@@ -81,3 +81,9 @@ extends:
     AdditionalModules:
       - name: perf-test-core
         groupId: com.azure
+    MatrixConfigs:
```
Comment (weshaggard) [inline sdk/core/ci.yml:84]: @benbp is this the best way to add this custom matrix. I don't like that we are duplicating the matrix and it would be great if we didn't need to do that.

Comment (benbp) [reply sdk/core/ci.yml:84]: @weshaggard I'm not a fan of the duplication either. We were trying to figure out a way to do this with the base matrix and were coming up empty, because if we do a high level `replace` operation for `TestGoals` then it will override the `TestGoals` entry when `TestFromSource=true` which is not desired. We don't have a "replace where xyz fields match" query type (and I believe that was an intentional decision to keep the replace syntax simple).

Another approach here could be to split `TestGoals` into two variables, one used in the base matrix, and one used in the includes (like `TestVariantGoals` or something), and then core could replace only `TestGoals`.

Comment (weshaggard) [reply sdk/core/ci.yml:84]: I feel we have to figure out a good way to share this or we definitely risk the core pipeline having different matrix configurations then the rest of the repo which isn't a good state to be in.
